---
title: "Home"
output_dir: "docs"
author: "Maria Ninova"
date: "April 21, 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Smt3 KD RNA-seq analysis with kallisto/sleuth

## Data pre-processing

The original FASTQ files (50 bp, single-end, reverse stranded) were processed with kallisto using a custom transcriptome, which is a combination of the REFSEQ mRNA sequences AND RepBase D. melanogaster TE seqeunces (refGene_plus_repbase).


> ./kallisto quant --single -t 4 -l 200 -s 50 -b 30 --rf-stranded -i transcriptome_plus_repbase/refGene_plus_repbase -o YL25 YL25_shW_RNAseq_1st.fq.gz

> ./kallisto quant --single -t 4 -l 200 -s 50 -b 30 --rf-stranded -i transcriptome_plus_repbase/refGene_plus_repbase -o YL26 YL26_shSmt3_RNAseq_1st.fq.gz

> ./kallisto quant --single -t 4 -l 200 -s 50 -b 30 --rf-stranded -i transcriptome_plus_repbase/refGene_plus_repbase -o YL27 YL27_shW_RNAseq_2nd.fq.gz

> ./kallisto quant --single -t 4 -l 200 -s 50 -b 30 --rf-stranded -i transcriptome_plus_repbase/refGene_plus_repbase -o YL28 YL28_shSmt3_RNAseq_2nd.fq.gz


Next, we use these files for sleuth analysis.

***
***

The refseq file is refGene.20150212.dm3.with_names.gtf.cleaned.fa, which is the one used in our standard pipeline fo eXpress analysis, and the repbase file is repbase_fly_20120418.fa. 
(origin /usr/local/home/manakov/Data/UCSC/dm3/refGene.20150212.dm3.with_names.gtf.cleaned.fa).
This is turn has been generated by 

> gffread -w $rawFasta -g refGene.20121022.dm3.with_names.gtf

And then this $rawFasta is further cleaned up by Sergei's script clean_fasta_from_gtf.pl which takes a refGene.gtf file, and returns a fasta file, where some duplicates are removed (see his description).
