---
title: "smt3 KD analysis genes aggredated by transcripts"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(sleuth)
library(dplyr)
library(stringr)
library(ggplot2)
library(DT)
#units<-"scaled_reads_per_base"
units<-"est_counts" # if doing transcript-centered.
```




## IMPORTANT NOTE

**This is transcript-level analysis.**
Importantly, it is also possible to use aggregation transcript p-values to perform gene differential expression, as introduced in [Yi et al., 2017](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1419-z). In this framework, differential expression is performed on transcripts as usual, but then transcript-level p-values can be aggregated to obtain gene-level p values.
The results are therefore different than the old method of gene_mode=T.
Also see [this walkthrough](https://pachterlab.github.io/sleuth_walkthroughs/pval_agg/analysis.html).

## Data pre-processing (always the same)

The original FASTQ files (50 bp, single-end, reverse stranded) were processed with kallisto using a custom transcriptome, which is a combination of the REFSEQ mRNA sequences AND RepBase D. melanogaster TE seqeunces (refGene_plus_repbase)

> ./kallisto quant --single -t 4 -l 200 -s 50 -b 30 --rf-stranded -i transcriptome_plus_repbase/refGene_plus_repbase -o YL25 YL25_shW_RNAseq_1st.fq.gz

> ./kallisto quant --single -t 4 -l 200 -s 50 -b 30 --rf-stranded -i transcriptome_plus_repbase/refGene_plus_repbase -o YL26 YL26_shSmt3_RNAseq_1st.fq.gz

> ./kallisto quant --single -t 4 -l 200 -s 50 -b 30 --rf-stranded -i transcriptome_plus_repbase/refGene_plus_repbase -o YL27 YL27_shW_RNAseq_2nd.fq.gz

> ./kallisto quant --single -t 4 -l 200 -s 50 -b 30 --rf-stranded -i transcriptome_plus_repbase/refGene_plus_repbase -o YL28 YL28_shSmt3_RNAseq_2nd.fq.gz

Preparing sleuth object

```{r create_so, warning=FALSE}

s2c<-data.frame(sample=c("YL25","YL27","YL26","YL28"),condition=c("ctrl", "ctrl","KD","KD"), 
                path=c("../kallisto/YL25/","../kallisto/YL27/", "../kallisto/YL26/","../kallisto/YL28/"))
s2c$path<-as.character(s2c$path)
print(s2c)

#this is to create a transcript to gene name dictionary - I just transform each fasta header.
t<-read.table("../kallisto/YL26/abundance.tsv", header=T, as.is=T)
t2g<-data.frame("target_id"=t$target_id, gene=sub(":CDS.*","",sub(".*gene=","", t$target_id)))
t2g$coding<-grepl("CDS", t2g$target_id)
t2g$TE<-grepl("\\|", t2g$target_id)

# number of elements of different type: with CDS annotation, RepBase entries (TE), and none (ncRNAs or genes with no CDS)
t2g %>% select(gene,coding, TE) %>% distinct() %>% group_by(coding, TE) %>% tally()

# make "so", note that aggregation column is supplied, but gene_mode is FALSE
so <- sleuth_prep(s2c, ~condition, extra_bootstrap_summary=T, read_bootstrap_tpm = TRUE, aggregation_column = 'gene', target_mapping = t2g, gene_mode=FALSE) 

```


***
***

#### Input values

Here, we can inspect the input data table. This is in transcripts as aggregation is for the p-values.


```{r, display_matrix, message=FALSE}
units

table1<-sleuth_to_matrix(so, which_df = "obs_norm", which_units = units) %>% as.data.frame() %>% 
  rename(shW.1=YL25, shW.2=YL27, shSmt3.1=YL26, shSmt3.2=YL28) 
table1 %>% round(digits=2) %>% datatable(options=list(pageLength=5))

```

***

## Analys–µs

Running sleuth Likelihood Ratio test (LRT), and displaying a table of significantly changed genes/tx, qval<0.05. This table DOES NOT have information about up- or down- regulation.

```{r sleuth_lrt,  message=FALSE}

so <- sleuth_fit(so, ~condition,'full')
so <- sleuth_fit(so, ~1, 'reduced')
so <- sleuth_lrt(so, 'reduced', 'full')

sleuth_results_LRT_gene <- sleuth_results(so, 'reduced:full', test_type = 'lrt', show_all = TRUE)
sleuth_results_LRT_tx <-sleuth_results(so, 'reduced:full', test_type = 'lrt', show_all = TRUE, pval_aggregate = F)

#gene aggregated table
dplyr::filter(sleuth_results_LRT_gene, qval <= 0.05) %>% datatable(options = list(pageLength=5))
#tx aggregated table
dplyr::filter(sleuth_results_LRT_tx, qval <= 0.05) %>% datatable(options = list(pageLength=5))

```

***
***

Running wald test & plotting MA plot. Note that there are beta-vals only for the tx-based table where the diff.expr was performed, therefore, the MA-plot is **transcript-based**.

*Note: the displayed table shows all results, however the MA plot does not show txs which did not pass filters, and Inf values.* 

```{r, sleuth_wald,  message=FALSE, fig.height=5, fig.width=6, warning=F}
so <- sleuth_wt(so, 'conditionKD')
sleuth_results_wald_gene <- sleuth_results(so, 'conditionKD', show_all = TRUE)
sleuth_results_wald_tx<- sleuth_results(so, 'conditionKD', show_all = TRUE, pval_aggregate = F)

#gene aggregated table
sleuth_results_wald_gene %>% datatable(options = list(pageLength=5))

#tx aggregated table
sleuth_results_wald_tx %>% datatable(options = list(pageLength=5))

# MA plot, transcript analysis. Adding count data for later.
table.counts<-sleuth_to_matrix(so, which_df = "obs_norm", which_units = units) %>% as.data.frame() %>% rename(shW.1=YL25, shW.2=YL27, shSmt3.1=YL26, shSmt3.2=YL28) %>% tibble::rownames_to_column("target_id") %>% mutate(x=(shW.1+shW.2)/2, y=(shSmt3.1+shSmt3.2)/2) %>% mutate(log2FC=log2(y/x))

res<-sleuth_results(so, 'conditionKD', show_all = FALSE, pval_aggregate = F) %>% left_join(table.counts, by="target_id") %>% mutate(signFC=(qval<0.05 & abs(log2FC)>1), significant=qval<0.05) %>% arrange(qval)

p<-ggplot(res, aes(mean_obs, b)) + theme_classic() + ggtitle("MA plot of transcripts and TEs, sleuth")
p <- p + xlab(paste("mean( log(counts+ 0.5 ) )"))
p <- p + ylab(paste0("beta: conditionKD")) 
p <- p + geom_text(data=(sleuth_results_wald_tx %>% filter(str_detect(target_id,"(smt3|^R1_DM|^R2_DM)"))), aes(mean_obs, b, label=sub(".*gene=(.*):CDS.*", "\\1",sub("_DM.*","",target_id))), position = position_nudge(x=0.5))

p1 <- p + geom_point(aes(colour = significant), alpha = 0.2) + scale_colour_manual(values = c("black", "green4"))
p1

```

Fig. MA plot from sleuth analysis, Wald test, transcript-based. All transcripts with qvalue<0.05 are highlighted (warning: beta != log2FC).

Below it is the same MA plot, but I only marked genes with >2 fold change, calculated from the original values in the sleuth table above.

```{r, sleuth_wald_FC,  message=FALSE, fig.height=5, fig.width=6}
p2 <- p + geom_point(aes(colour = signFC), alpha = 0.2) + scale_colour_manual(values = c("black", "red"))
p2
```

Fig. MA plot from sleuth analysis, transcript-level, Wald test. (warning: beta != log2FC). Only genes with qval<0.05 & log2FC (from kallisto numbers) are highlighted.

***
***

### Alternative visualization

This is an alternative visualization of the data, where we do a scatter plot in control vs knockdown.
We can choose to select the dots based on **qvalues either** from the **Wald test**, or the **LTR test**. The test choice does not change the interpretation of the data in the light of the R1, R2 up-regulation compared to genes. 

```{r scatter, fig.width=15, fig.height=6}
#values tables
table.tpm<-sleuth_to_matrix(so, which_df = "obs_norm", which_units = "tpm") %>% as.data.frame() %>% rename(shW.1=YL25, shW.2=YL27, shSmt3.1=YL26, shSmt3.2=YL28) %>% tibble::rownames_to_column("target_id") %>% mutate(x=(shW.1+shW.2)/2, y=(shSmt3.1+shSmt3.2)/2) %>% mutate(log2FC=log2(y/x))

tab.raw<-full_join(table.counts, table.tpm, by="target_id",suffix=c(".count",".tpm")) 

### stats tables
tab.wald<-sleuth_results(so, 'conditionKD', show_all = TRUE, pval_aggregate = F) %>% arrange(target_id) # wald test
tab.lrt<-sleuth_results(so, 'reduced:full', test_type = 'lrt', show_all = TRUE, pval_aggregate = F) %>% arrange(target_id) #lrt test
tab.stat<-full_join(tab.wald, tab.lrt, by="target_id", suffix=c(".wald",".lrt")) 

### all data together, fold change is taken from the est_counts 
tab.plot<-full_join(tab.stat, tab.raw, by="target_id") %>% mutate(signWald=(abs(log2FC.count)>1 & qval.wald<0.05), signLRT=(abs(log2FC.count)>1 & qval.lrt<0.05))

### plot
p.sub<-geom_text(data=(tab.plot %>% filter(str_detect(target_id,"(smt3|^R1_DM|^R2_DM)"))), aes(log2(x.tpm+1), log2(y.tpm+1), label=sub(".*gene=(.*):CDS.*", "\\1",sub("_DM.*","",target_id))), position = position_nudge(x=1))

top20.x<-tail(sort(tab.plot$x.tpm),20)[1]
top20.y<-tail(sort(tab.plot$y.tpm),20)[1]

p<-ggplot(tab.plot, aes(log2(x.tpm+1), log2(y.tpm+1))) + theme_classic() + theme(legend.position="left") + p.sub
p<-p + scale_color_manual(values=c("black", "beige", "green4")) 
p<-p + xlab(paste("shW, log2 (av.tpm + 1) ")) + ylab(paste("shSmt3, log2 (av.tpm + 1) "))
p<-p + geom_hline(yintercept = log2(top20.y+1), linetype="dotted") + geom_vline(xintercept = log2(top20.x+1), linetype="dotted") 

p1<-p + geom_point(aes(colour = paste(signWald)), alpha = 0.4) + geom_abline(intercept = 0, slope=1, col="blue")  
p2<-p + geom_point(aes(colour = paste(signLRT)), alpha = 0.4)  + geom_abline(intercept = 0, slope=1, col="blue")  

p3<-ggExtra::ggMarginal(p1, type="histogram", size=3, fill = "grey70")
p4<-ggExtra::ggMarginal(p2, type="histogram", size=3, fill = "grey70")

gridExtra::grid.arrange(p3, p4, nrow=1)

```

Figure. Scatter plots of transcript expression in control vs KD. The dotted line denotes area that contains the top 20 most highly expressed genes in the corresponding condition. Values log2FC>2 and significant: qvalue (Wald or LRT) < 0.05 are highlighted.
The histograms above the y-axis and x-axis show the distribution of values. Blue line shows 1:1 ratio.

*Note: Here, pale dots show values ("NA") which did not pass various sleuth filters and should be ignored. Can be removed by "show_all=FALSE".*

***
***

## Numerical summary, **transcripts**

First we prepare the data in a convenient table of all.

```{r data_prep}
tab.all<-tab.plot %>% mutate("filt_out"=is.na(b), "type"=as.factor(ifelse(str_detect(target_id, "\\|"),"TE","RefSeq")))
```

***

Number of RepBase (TE) and other (protein-coding genes,ncRNA, etc) entries from RefSeq (all genes) analyzed; some were filtered out because they didn't pass sleuth criteria (usually very low read counts or large discrepancy between reps).

```{r num_summary}
tab.all %>% group_by(type, filt_out) %>% tally() %>% arrange(filt_out)
```

***


Number of significantly changed transcripts (qval 0.05, LRT or Walt Test; logFC (onscaled_reads_per_base)):

```{r signFC_by_type}
#Wald test
tab.all %>% filter(abs(log2FC.count)>1, qval.wald<0.05) %>% mutate("change"=ifelse(log2FC.count>0,"up", "down")) %>% group_by(type,change) %>% tally() %>% arrange(desc(change))
#LRT
tab.all %>% filter(abs(log2FC.count)>1, qval.lrt<0.05) %>% mutate("change"=ifelse(log2FC.count>0,"up", "down")) %>% group_by(type,change) %>% tally() %>% arrange(desc(change))

```

#### Summary of the numbers above:
From 17975 RefSeq transcripts and 166 RebBase entries which passed the initial filters, 
187(56) and 36(23) are significantly up-regulated more than 2-fold in smt3 KD,
and 159(64) or 2 are down-regulated more than 2-fold (qvalue 0.05, Wald test or LTR test).

**Therefore, smt3 KD led to >2 fold up/down -regulation for ~1% of the transcripts (protein-coding and ncRNAs), and up-regulation of ~20% of the TEs.**

**R1 and R2 - from relatively lowly expressed genes in the control - become among the top 20 most abundant products in SUMO KD.**
See below:

***

Although many txs change, it is evident from the plots mostly modestly expressed ones show large FC, except for R1 and R2. If we filter txs with over 100 times change in expression log2FC>=~6.64, significant in at least one test:

```{r FC_filter, warning=FALSE, message=F}
tab.all %>% filter(log2FC.count>log2(100)) %>% filter(qval.wald<0.05 | qval.lrt<0.05) %>%
                    select(target_id, shW.1.count, shW.2.count, shSmt3.1.count, shSmt3.2.count, log2FC.count, type, b, qval.wald, qval.lrt) %>% mutate_at(2:5, funs(round(.,3))) %>% arrange(-( shSmt3.1.count*0.5+shSmt3.2.count*0.5)) %>% datatable(options=list(pageLength=5))
```

Results show that all other genes, apart from R1 and R2 in this set, are ones that practically have no reads in the control, and just few reads in the KD. 

***

Numbers below show the expression of R1 and R2 elements; including rank of the average values. "X" and "Y" are average per control and KD, respectively:

*Note: Values can be seen as transcripts per million (TPM), see [this explanation of TPM]( https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/)
, or just counts.

```{r Rs_expr_rank}
tab.all.ranks1<-tab.all %>% filter(!filt_out) %>% select(target_id, matches("sh.*count"),matches("(x|y).count")) %>% mutate(ctrl.rank.count=rank(desc(x.count)), kd.rank.count=rank(desc(y.count)))

tab.all.ranks2<-tab.all %>% filter(!filt_out) %>% select(target_id, matches("sh.*tpm"),matches("(x|y).tpm")) %>% mutate(ctrl.rank.tpm=rank(desc(x.tpm)), kd.rank.tpm=rank(desc(y.tpm)))

tab.all.ranks1 %>% filter(str_detect(target_id, "^R[12]_"))
tab.all.ranks2 %>% filter(str_detect(target_id, "^R[12]_"))
```

**Notably, R1 and R2 change from relatively lowly expressed elements, to the top 20 most abundant transcripts (in TPM). No other elements shows such a dramatic increase.**


***
***


## **Gene-based** analysis with **p-value aggregation**

